File: DS/Linkedlist/singly_linked_list/PYTHON/insert_node.py
class Node:
    def __init__(self, num):
        self.data = num
        self.link = None


class LinkedList:

    def __init__(self):
        self.head = None

    # insert node at the end of linked list
    def append(self, num):
        temp = Node(num)
        if self.head is None:
            self.head = temp
            return

        temp1 = self.head
        while temp1.link is not None:
            temp1 = temp1.link

        temp1.link = temp

    # insert node at the front of linked list
    def insert_head(self, num):
        temp = Node(num)
        if self.head is None:
            head = temp
            return

        temp.link = self.head
        self.head = temp
    
    # insert a node at given position
    def insert_nth(self, loc, num):
        temp = Node(num)
        if loc == 1:
            temp.link = self.head
            self.head = temp
            return
        loc -= 2

        temp1 = self.head
        while loc:
            temp1 = temp1.link
            loc -= 1

        temp.link = temp1.link
        temp1.link = temp
    
    # print
    def print_list(self):
        temp = self.head
        while temp:
            print(temp.data, end=" ")
            temp = temp.link
            
if __name__ == '__main__':

    l_list = LinkedList()
    
    l_list.append(1)
    l_list.append(2)
    l_list.append(3)
    l_list.append(4)
    l_list.append(5)
    l_list.insert_head(6)
    l_list.insert_nth(3, 10)
    
    l_list.print_list()
    
    

File: DS/Linkedlist/singly_linked_list/C/insert_node.c
#include <stdio.h>
#include <stdlib.h>

struct Node{
    int data;
    struct Node *link;
};

struct Node *head;//pointer to head node

// insert node at the end of linked list
void insert_end(int num){
    struct Node *temp = (struct Node*)malloc(sizeof(struct Node));
    temp->data = num;
    if(head == NULL){
        temp->link = NULL;
        head = temp;
    }

    else{
        temp->link = NULL;
        struct Node *temp1 = head;
        while (temp1->link!=NULL)
        {
            temp1 = temp1->link;
        }

        temp1->link = temp;  
    } 
}

// insert a node at the front of linked list

void insert_head(int num){
    struct Node *temp = (struct Node*)malloc(sizeof(struct Node));
    temp->data = num;
    if(head == NULL){
        temp->link = NULL;
    }

    else{
        temp->link = head;
    }
    head = temp;
}

// insert a node at given position

void insert_nth(int num,int loc){
    struct Node *temp = (struct Node*)malloc(sizeof(struct Node));
    temp->data = num;

    if(loc == 1){
        temp->link = NULL;
        head = temp;
    }

    else{
        loc -= 2;
        struct Node *temp1 = head;
        while (loc--)
        {
            temp1 = temp1->link;
        }
        temp->link = temp1->link;
        temp1->link = temp; 
    }
}

// print

void print_list(){
    struct Node *temp = head;
    while (temp!=NULL)
    {
        printf("%d ", temp->data);
        temp = temp->link;
    }
}

// main

int main(){
    int size;
    scanf("%d", &size);

    head = NULL;

    while (size--)
    {
        int num;
        scanf("%d", &num);
        insert_end(num);
    }

    insert_head(40);

    insert_nth(50,3);

    print_list();
}
